### PR TITLE
COM-908 Add test contact list

### PR DESCRIPTION
**Pull request by juliawegmayr**
-   [x] Add changeset (if necessary)

- Create a test contact list
- Fix: if a contact is added to the main list, it was added to all lists without filters. Added a check for manuallyAssignedContacts.
- Contacts added to a test list, don't need a double opt-in.
- If a contact is added to the test list, it is possible to add it to the main list as well. Then a DOI is sent.
- Test list is not displayed in the target group page.

### CODE REVIEW

File: demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx [hunk 1]
```diff
@@ -110,10 +108,8 @@ export const getBrevoContactConfig = (
...
             </>
         ),
-        input2State: (values?: AdditionalFormConfigInputProps) => {
+        input2State: (values: AdditionalFormConfigInputProps) => {
             return {
-                email: values?.email ?? "",
-                redirectionUrl: values?.redirectionUrl ?? "",
                 attributes: {
                     SALUTATION: values?.attributes?.SALUTATION,
                     FIRSTNAME: values?.attributes?.FIRSTNAME,
```
Comment (RainbowBunchie) [inline demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx:115]: Since the values are not nullable anymore, you can remove the '?' here as well

File: packages/admin/src/brevoContacts/form/BrevoContactForm.tsx [hunk 2]
```diff
@@ -42,9 +42,13 @@ import {
 } from "./BrevoContactForm.gql.generated";
 
 export type EditBrevoContactFormValues = {
+    [key: string]: unknown;
+};
+
+type BaseBrevoContactFormValues = EditBrevoContactFormValues & {
+    //Umbenennen
```
Comment (RainbowBunchie) [inline packages/admin/src/brevoContacts/form/BrevoContactForm.tsx:49]: Should hits still be renamed? Or can the comment be deleted?

File: packages/admin/src/brevoContacts/form/BrevoContactForm.tsx [hunk 3]
```diff
@@ -76,11 +77,15 @@ export function BrevoContactForm({ id, scope, input2State, additionalFormFields,
...
 
-    const initialValues = React.useMemo<Partial<EditBrevoContactFormValues>>(() => {
+    const initialValues = React.useMemo<Partial<BaseBrevoContactFormValues>>(() => {
         let additionalInitialValues = {};
 
         if (input2State) {
-            additionalInitialValues = input2State({ email: "", redirectionUrl: "", ...data?.brevoContact });
+            additionalInitialValues = input2State({
+                email: "",
+                redirectionUrl: "",
```
Comment (RainbowBunchie) [inline packages/admin/src/brevoContacts/form/BrevoContactForm.tsx:86]: these values should be in the `additionalInitialValues` (would rename to `baseInitalValues`). Since they are not needed in the `input2State` anymore. And the `input2state` values should be added to these `baseInitalValues`

File: packages/api/src/mikro-orm/migrations/Migration20240830112400.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { Migration } from "@mikro-orm/migrations";
+
+export class Migration20240830112400 extends Migration {
+    async up(): Promise<void> {
+        this.addSql('alter table "TargetGroup" add column "isTestList" boolean null;');
```
Comment (RainbowBunchie) [inline packages/api/src/mikro-orm/migrations/Migration20240830112400.ts:5]: This should not be nullable. You can set existing values to `false` since there are no existing test lists because it is a new feature.

File: packages/api/src/brevo-contact/brevo-contacts.service.ts [hunk 5]
```diff
@@ -53,12 +53,31 @@ export class BrevoContactsService {
...
+    }
+
     public async getTargetGroupIdsForNewContact({
         contactAttributes,
         scope,
+        assignedContactsTargetGroupBrevoId,
     }: {
         contactAttributes?: BrevoContactAttributesInterface;
         scope?: EmailCampaignScopeInterface;
+        assignedContactsTargetGroupBrevoId?: number;
```
Comment (RainbowBunchie) [inline packages/api/src/brevo-contact/brevo-contacts.service.ts:80]: This is not used, please remove

File: packages/api/src/brevo-contact/brevo-contacts.service.ts [hunk 6]
```diff
@@ -53,12 +53,31 @@ export class BrevoContactsService {
...
+        attributes,
+        scope,
+    }: {
+        email: string;
+        attributes?: BrevoContactAttributesInterface;
+        scope: EmailCampaignScopeInterface;
+    }): Promise<boolean> {
+        const testTargetGroupForScope = await this.targetGroupService.createIfNotExistTestTargetGroupForScope(scope);
+
+        const brevoIds = [testTargetGroupForScope.brevoId];
```
Comment (RainbowBunchie) [inline packages/api/src/brevo-contact/brevo-contacts.service.ts:67]: imo you do not need this variable, can simply pass the array like this in the brevo `createTestContact` function.

File: packages/api/src/brevo-contact/brevo-contact.resolver.ts [hunk 7]
```diff
@@ -92,6 +95,39 @@ export function createBrevoContactResolver({
             return new PaginatedBrevoContacts(contacts, count, { offset, limit });
         }
 
+        @Query(() => PaginatedBrevoContacts)
+        async brevoTestContacts(@Args() { offset, limit, email, targetGroupId, scope }: BrevoContactsArgs): Promise<PaginatedBrevoContacts> {
```
Comment (RainbowBunchie) [inline packages/api/src/brevo-contact/brevo-contact.resolver.ts:99]: You can remove the `targetGroupId` here -> this query can only be used to query the test list. if you set a   `targetGroupId` in the args that is not the test list id this will not return anything anyways

**Review by RainbowBunchie — CHANGES_REQUESTED** [review]

File: demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx (hunk 1 above)
Comment (juliawegmayr) [reply demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx:115]: If I do so, I get the error:
_Cannot read properties of undefined (reading 'attributes')_, when I open the BrevoContactForm 🤔 

Did we change anything for those values? I think, they still can be undefined

Comment (RainbowBunchie) [reply demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx:115]: In that case the `?` should be readded in the function params in line 111. I do not know why this was removed 🤷🏻‍♀️

File: packages/api/src/mikro-orm/migrations/Migration20240830112400.ts (hunk 4 above)
Comment (RainbowBunchie) [reply packages/api/src/mikro-orm/migrations/Migration20240830112400.ts]: Drop the default value after setting all rows to false 🙏🏻

File: packages/admin/src/brevoContacts/form/BrevoContactForm.tsx (hunk 2 above)
Comment (RainbowBunchie) [reply packages/admin/src/brevoContacts/form/BrevoContactForm.tsx]: Now that I think about it: I do not think `Base` is not really the correct name for this. Since this extends the BaseFormValues (defined in `& { ...}`} with the configured values. Something like `BrevoContactFormValuesWithAttributes` would be easier to understand I think. What do you think?

Comment (juliawegmayr) [reply packages/admin/src/brevoContacts/form/BrevoContactForm.tsx]: Maybe `EditBrevoContactFormValuesWithAttributes`?

Comment (RainbowBunchie) [reply packages/admin/src/brevoContacts/form/BrevoContactForm.tsx]: 👍🏻

File: packages/api/src/brevo-contact/brevo-contact.resolver.ts [hunk 8]
```diff
@@ -184,13 +210,103 @@ export function createBrevoContactResolver({
...
+        @Mutation(() => Boolean)
+        @AffectedEntity(BrevoContact)
+        async deleteBrevoTestContact(
+            @Args("id", { type: () => Int }) id: number,
+            @Args("scope", { type: () => Scope }, new DynamicDtoValidationPipe(Scope)) scope: typeof Scope,
+        ): Promise<boolean> {
+            const contact = await this.brevoContactsApiService.findContact(id, scope);
+            const where: FilterQuery<TargetGroupInterface> = { scope, isMainList: false, isTestList: true };
+            const testTargetGroup = await this.targetGroupRepository.findOne(where);
+            const contactIncludesMainList = testTargetGroup?.brevoId ? contact.listIds.includes(testTargetGroup.brevoId) : false;
```
Comment (raphaelblum) [inline packages/api/src/brevo-contact/brevo-contact.resolver.ts:291]: why `contactIncludesMainList`?
shouldn't that be `contactIncludesTestList` ?

File: packages/api/src/brevo-api/brevo-api-contact.service.ts [hunk 9]
```diff
@@ -95,8 +110,8 @@ export class BrevoApiContactsService {
             if (!contact) return undefined;
             return contact;
         } catch (error) {
-            // Brevo throws 400 error if no contact was found
-            if (isErrorFromBrevo(error) && error.response.statusCode === 400) {
+            // Brevo throws 404 error if no contact was found
+            if (isErrorFromBrevo(error) && error.response.statusCode === 404) {
```
Comment (raphaelblum) [inline packages/api/src/brevo-api/brevo-api-contact.service.ts:114]: why is this changed? In the main branch thomas changed that to catch both errors:
https://github.com/vivid-planet/comet-brevo-module/pull/94/files

Comment (juliawegmayr) [reply packages/api/src/brevo-api/brevo-api-contact.service.ts]: I had problems, as I got a 404 error and I thought that the wrong code was given in the code. I checked it out at Brevo now, and I agree, that both codes should be handled. Thanks for fixing it and pointing it out here as well. 👍

File: packages/api/src/target-group/target-groups.service.ts [hunk 10]
```diff
@@ -141,6 +139,27 @@ export class TargetGroupsService {
...
     }
 
+    public async createIfNotExistTestTargetGroupForScope(scope: EmailCampaignScopeInterface): Promise<TargetGroupInterface> {
+        const testList = await this.repository.findOne({ scope, isMainList: false, isTestList: true });
+
+        if (testList) {
+            return testList;
+        }
+
+        const title = "Test list for current scope";
```
[the remark this answers is not in the document]
Comment (RainbowBunchie) [reply packages/api/src/target-group/target-groups.service.ts]: I guess just stringifying the scope would be enough here? @raphaelblum

[the remark this answers is not in the document]
Comment (juliawegmayr) [reply packages/api/src/target-group/target-groups.service.ts]: Thank you :)
I didn't think about it

File: packages/api/src/mikro-orm/migrations/Migration20240830112400.ts
Comment (juliawegmayr) [inline packages/api/src/mikro-orm/migrations/Migration20240830112400.ts:1]: @RainbowBunchie 
When I checked out the branch, I realised, that I get an error, when creating target group lists, stating that testList was not found. I am not sure, if this is correct, but I deleted the drop default again, as it resolved all errors.

Comment (RainbowBunchie) [reply packages/api/src/mikro-orm/migrations/Migration20240830112400.ts:1]: the dropping of the default should be included in the migration. what exactly was the error here? You can also set a default value in typescript for the `isTestList`, so the default value is apparent when looking in the code. The problem with database default values is that they are really hard to trace.

File: packages/api/src/target-group/target-groups.service.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (raphaelblum) [reply packages/api/src/target-group/target-groups.service.ts]: I think it would still be better to make it more user-readable. 
But we can think of improving this in another task, since we also should update it for the main list

File: .changeset/chilly-dogs-draw.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+---
+"@comet/brevo-admin": major
+"@comet/brevo-api": major
+---
+
+Adds `createBrevoTestContactsPage` for creating a test contacts page, that is indepent from the main list.
+
+Remove `email` and `redirectionUrl` from `brevoContactsPageAttributesConfig`
```
Comment (raphaelblum) [inline .changeset/chilly-dogs-draw.md:8]: Couldn't this be a minor version?

There are no breaking changes as far as I can see. It's a new feature with the test list, but it must not be used as far as I can see.

File: packages/api/src/mikro-orm/migrations/Migration20240830112400.ts
Comment (raphaelblum) [reply packages/api/src/mikro-orm/migrations/Migration20240830112400.ts:1]: Yes we should remove the default again. There was probably another error.  You could check out the main branch and execute the fixtures, add some data and then switch back to this branch and check if the error is still there.

We can also take a look at the error together if it occurs again

File: packages/api/src/brevo-contact/dto/brevo-test-contact-input.factory.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        })
+        class BrevoTestContactInputBase implements BrevoTestContactInputInterface {
+            @IsNotEmpty()
+            @IsString()
+            @Field()
+            email: string;
+
+            @IsBoolean()
+            @Field()
+            @IsOptional()
```
Comment (raphaelblum) [inline packages/api/src/brevo-contact/dto/brevo-test-contact-input.factory.ts:39]: if possible, please rather use `IsUndefinable` or `IsNullable`

(there might be a lot of places where this is not the case yet, but when starting the project IsUndefinable and IsNullable where not yet available)

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx [hunk 13]
```diff
@@ -263,12 +264,12 @@ export function TargetGroupsGrid({
...
     ];
 
-    const { filter: gqlFilter, search: gqlSearch } = muiGridFilterToGql(columns, dataGridProps.filterModel);
+    const { search: gqlSearch } = muiGridFilterToGql(columns, dataGridProps.filterModel);
 
     const { data, loading, error } = useQuery<GQLTargetGroupsGridQuery, GQLTargetGroupsGridQueryVariables>(targetGroupsQuery, {
         variables: {
             scope,
-            filter: gqlFilter,
+            filter: { isTestList: { equal: false || null } },
```
Comment (raphaelblum) [inline packages/admin/src/targetGroups/TargetGroupsGrid.tsx:272]: isn't this:
`filter: { isTestList: { equal: false || null } },`

the same as this:
`filter: { isTestList: { equal: null } },`

File: .changeset/chilly-dogs-draw.md (hunk 11 above)
Comment (juliawegmayr) [reply .changeset/chilly-dogs-draw.md:8]: True, I initially thought, that the change in the `brevoContactsPageAttributesConfig` breaks the contacts page and the create/update functions. But that's not true, so minor should be fine 👍

File: packages/api/src/mikro-orm/migrations/Migration20240830112400.ts
Comment (juliawegmayr) [reply packages/api/src/mikro-orm/migrations/Migration20240830112400.ts:1]: I'm so sorry, I tested it again and it worked perfectly fine. I guess, I had the problems because of switching branches. 
Sorry for the confusion and thanks for the clarification!

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx (hunk 13 above)
Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/TargetGroupsGrid.tsx]: `isTestList` is not nullable though

File: .changeset/chilly-dogs-draw.md (hunk 11 above)
Comment (RainbowBunchie) [reply .changeset/chilly-dogs-draw.md]: Is the change in `EditBrevoContactFormValues` not a breaking change? @raphaelblum @juliawegmayr 
just to be sure, i thought this is breaking since before the email and redirectionUrl had to be included in the `createBrevoContactPage` and now they should not be present there. Or am I wrong here?
 
see changes in `demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx`

```
export type EditBrevoContactFormValues = {
    [key: string]: unknown;
    // email: string; got removed 
    // redirectionUrl: string;  got removed
};
```

Comment (raphaelblum) [reply .changeset/chilly-dogs-draw.md]: It is still possible to pass the email and redirectionUrl, or is it not?

File: packages/api/src/brevo-contact/brevo-contact.resolver.ts [hunk 14]
```diff
@@ -184,13 +210,103 @@ export function createBrevoContactResolver({
...
+                    {
+                        listIds: linkListIds,
+                        unlinkListIds: [testListId],
+                    },
+                    scope,
+                );
+                return true;
+            } else {
+                return this.brevoContactsApiService.deleteContact(id, scope);
+            }
```
Comment (raphaelblum) [inline packages/api/src/brevo-contact/brevo-contact.resolver.ts:309]: What exactly should happen here?

**Am I right with that:**
When the contact is only in the testlist, and nowhere else, it should be deleted.
Otherwise it should just be unlinked

**Currently it works like this:**
 - When there is a testTargetGroup AND the contact is in the list, it will be unlinked, otherwise it will be deleted? Does this make sense?

File: .changeset/chilly-dogs-draw.md (hunk 11 above)
Comment (juliawegmayr) [reply .changeset/chilly-dogs-draw.md]: As long as the test contacts are not used, it is still possible to pass email and redirectionUrl.
I also tested it by using the initial config, like it is now on the main branch.

Comment (RainbowBunchie) [reply .changeset/chilly-dogs-draw.md]: I thought this would result in an error. If not, then there is no problem 👍🏻  sorry for the confusion

Comment (raphaelblum) [reply .changeset/chilly-dogs-draw.md]: > As long as the test contacts are not used, it is still possible to pass email and redirectionUrl.

Okay good then it is non-breaking. The target branch can be switched to main as @RainbowBunchie said.

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx (hunk 13 above)
Comment (juliawegmayr) [reply packages/admin/src/targetGroups/TargetGroupsGrid.tsx]: MainList is not nullable. 
I checked the code again and adapted it, so that isMainList and isTestList are both not nullable.

File: packages/api/src/brevo-contact/brevo-contact.resolver.ts (hunk 14 above)
Comment (juliawegmayr) [reply packages/api/src/brevo-contact/brevo-contact.resolver.ts]: Yes, you are absolutely right! 
Now I check, if the contact is in the MainList and if he is, the test list is unlinked. Otherwise deleted

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx (hunk 13 above)
Comment (juliawegmayr) [reply packages/admin/src/targetGroups/TargetGroupsGrid.tsx]: c9a0e639675b628a572cb301f2df744e1cee88d8

**Review by raphaelblum — DISMISSED** [review]

**Review by raphaelblum — APPROVED** [review]

File: demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx [hunk 15]
```diff
@@ -70,9 +70,7 @@ export const getBrevoContactConfig = (
         name: string;
     };
     input2State: (values?: AdditionalFormConfigInputProps) => {
-        email: string;
-        redirectionUrl: string;
-        attributes: { BRANCH?: Array<GQLBrevoContactBranch>; SALUTATION?: GQLBrevoContactSalutation; FIRSTNAME?: string; LASTNAME?: string };
+        attributes: { SALUTATION?: GQLBrevoContactSalutation; FIRSTNAME?: string; LASTNAME?: string };
```
Comment (RainbowBunchie) [inline demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx:73]: Why did the branch get removed here?

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx [hunk 16]
```diff
@@ -50,6 +50,7 @@ const targetGroupsFragment = gql`
         totalSubscribers
         totalContactsBlocked
         isMainList
+        isTestList
```
Comment (RainbowBunchie) [inline packages/admin/src/targetGroups/TargetGroupsGrid.tsx:53]: This is not used in the frontend, can this be removed

Comment (RainbowBunchie) [reply packages/admin/src/targetGroups/TargetGroupsGrid.tsx:53]: will be false for every target group anyways since it gets filtered in the api

**Review by RainbowBunchie — CHANGES_REQUESTED** [review]

File: demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx (hunk 15 above)
Comment (juliawegmayr) [reply demo/admin/src/common/brevoModuleConfig/brevoContactsPageAttributesConfig.tsx]: c9a0e639675b628a572cb301f2df744e1cee88d8

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx (hunk 16 above)
Comment (juliawegmayr) [reply packages/admin/src/targetGroups/TargetGroupsGrid.tsx]: 51c898ce3c997e11ef38c54df1ba856b202a056c

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx [hunk 17]
```diff
@@ -263,12 +263,11 @@ export function TargetGroupsGrid({
         },
     ];
 
-    const { filter: gqlFilter, search: gqlSearch } = muiGridFilterToGql(columns, dataGridProps.filterModel);
+    const { search: gqlSearch } = muiGridFilterToGql(columns, dataGridProps.filterModel);
```
Comment (RainbowBunchie) [inline packages/admin/src/targetGroups/TargetGroupsGrid.tsx:266]: why did the filters get removed here?
the targetgroup grid has filters right?

**Comment by RainbowBunchie** [comment]
@juliawegmayr I just checked out the branch and the test list shows up in the target groups 
Can you please take a look at this? 
Screenshot 2024-10-25 at 12 55 17

**Review by RainbowBunchie — CHANGES_REQUESTED** [review]

File: packages/admin/src/targetGroups/TargetGroupsGrid.tsx (hunk 17 above)
Comment (juliawegmayr) [reply packages/admin/src/targetGroups/TargetGroupsGrid.tsx]: Initially I wanted to filter for the test contact list here, not in the api. After undoing my changes, I forgot to add the filters here again. I'm sorry! 
I corrected it here: fd436ca3cf5abca28f238d5af16f035623c85606

**Comment by juliawegmayr** [comment]
> @juliawegmayr I just checked out the branch and the test list shows up in the target groups Can you please take a look at this? Screenshot 2024-10-25 at 12 55 17

Fixed here: 205b281f06e28621620ef15f4c9707978add1d13

**Comment by RainbowBunchie** [comment]
> > @juliawegmayr I just checked out the branch and the test list shows up in the target groups Can you please take a look at this? Screenshot 2024-10-25 at 12 55 17
> 
> Fixed here: [205b281](https://github.com/vivid-planet/comet-brevo-module/commit/205b281f06e28621620ef15f4c9707978add1d13)

@juliawegmayr Shouldn't this work with the "normal" filters above? in "options.filter". Why do you need the custom filter logic here?

**Comment by juliawegmayr** [comment]
> > > @juliawegmayr I just checked out the branch and the test list shows up in the target groups Can you please take a look at this? Screenshot 2024-10-25 at 12 55 17
> > 
> > 
> > Fixed here: [205b281](https://github.com/vivid-planet/comet-brevo-module/commit/205b281f06e28621620ef15f4c9707978add1d13)
> 
> @juliawegmayr Shouldn't this work with the "normal" filters above? in "options.filter". Why do you need the custom filter logic here?

True! I removed my changes again: ebfe653f7989bb2afa644f54a2f36bd0a76f7873
Sorry for my missunderstandings.

**Review by raphaelblum — DISMISSED** [review]

**Comment by RainbowBunchie** [comment]
@juliawegmayr I just checked this out and I can still see the test group in the target groups. Did you test this? 
I would suggest you set `isTestList` to `false` in the targetGroups query in the api.

**Comment by juliawegmayr** [comment]
> @juliawegmayr I just checked this out and I can still see the test group in the target groups. Did you test this? I would suggest you set `isTestList` to `false` in the targetGroups query in the api.


I guess I didn't restart the api while I tested it and of course it seemed to work then. Totally my bad. I updated the code here:

4a9af5d1a41fcee8cf614f32daba016238640eaa

**Review by raphaelblum — APPROVED** [review]
